packages/dds/merge-tree/src/sortedSet.ts

```diff
@@ -0,0 +1,65 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+export abstract class SortedSet<T, U extends string | number> {
+    protected abstract getKey(t: T): U;
+
+    protected readonly keySortedItems: T[] = [];
+
+    public get size(): number {
+        return this.keySortedItems.length;
+    }
+
+    public get items(): readonly T[] {
+        return this.keySortedItems;
+    }
+
+    public addOrUpdate(newItem: T, update?: (existingItem: T, newItem: T) => void) {
+        const position = this.findOrdinalPosition(this.getKey(newItem));
+        if (position.exists) {
+            if (update) {
+                update(this.keySortedItems[position.index], newItem);
+            }
+        } else {
+            this.keySortedItems.splice(position.index, 0, newItem);
+        }
+    }
+
+    public remove(item: T): boolean {
+        const position = this.findOrdinalPosition(this.getKey(item));
+        if (position.exists) {
+            this.keySortedItems.splice(position.index, 1);
+            return true;
+        }
+        return false;
+    }
+
+    public has(item: T): boolean {
+        const position = this.findOrdinalPosition(this.getKey(item));
+        return position.exists;
+    }
+
+    private findOrdinalPosition(ordinal: U, start?: number, end?: number): { exists: boolean; index: number; } {
```

Comment:
there are also some fairly significant changes to this in next. might be best to make this change there, to avoid complex merges

Reply:
Thanks for calling this out! I've rebased my changes onto `next`.